### PR TITLE
 Fix for MASK_OUTSIDE_OBCS with MASKING_DEPTH

### PR DESCRIPTION
**Pull request by kshedstrom**
The MASK_OUTSIDE_OBCS flag doesn't know about the MASKING_DEPTH and this should take care of the problem.

### CODE REVIEW

**Review by Hallberg-NOAA — APPROVED** [review]
These changes mirror the code setting the `MASKING_DEPTH` elsewhere in the MOM6 code, and they make sense to me to add them here as well.

**Comment by kshedstrom** [comment]
I am hopefully done changing things on this branch now.

**Comment by adcroft** [comment]
@kshedstrom  In commit 01b0dc4ed you updated the way to handle the v-direction which I think makes sense. However, you didn't change the way the u-direction is handled and left it in the form of the previous commit. I don't think this explains the MacOS fails (which @marshallward suspects is a new gnu-compiler options thing) but currently I think it breaks the rotational symmetry rule.

**Review by adcroft — CHANGES_REQUESTED** [review]
@Hallberg-NOAA earlier approved the first version, and I agree with Bob that this seems like a good fix. However, a subsequent "better" commit broke symmetry (see https://github.com/NOAA-GFDL/MOM6/pull/752#issuecomment-2468396163). This seems easy to fix (apply to u- what was done to v-) but in the mean-time I'll mark this as "changes requested".

**Comment by kshedstrom** [comment]
The symmetric version worked for EW boundaries, but not NS boundaries. The algorithm is sweeping across lines of i and is inherently non-symmetric. With non-zero turns, it would need fixing, I suppose. I'll have to investigate how the turns are actually done.

**Comment by kshedstrom** [comment]
I rotated the Supercritical test and Thomas's test and I stand by my current version. The failure is:
```
FAIL: Diagnostics tc3.regression.diag have changed.
```
One version of Thomas's test was spinning up a circulating flow while the current version has zero flow with nothing spinning it up. The previous code didn't have eta zeroed out outside the OBCs. I wouldn't be surprised if those eta values were being used somehow. Do we need to set them to some wacky value and see what happens?

**Comment by kshedstrom** [comment]
The Thomas test when rotated runs when compiled for debugging, but not when compiled for repro. It fails with:
```
[chinook04:2911387:0:2911387] Caught signal 11 (Segmentation fault: address not mapped to object at address (nil))
==== backtrace (tid:2911387) ====
 0 0x0000000000012d20 __funlockfile()  :0  
 1 0x0000000001183528 __mpp_domains_mod_MOD_mpp_update_domain2d_r8_3dv()  //import/c1/AKWATERS/kate/ESMG/ESMG-configs/src/FMS2/mpp/include/mpp_update_domains2D.fh:518
 2 0x0000000000a69d5e __mom_domain_infra_MOD_pass_vector_3d()  //import/c1/AKWATERS/kate/ESMG/ESMG-configs/src/MOM6/config_src/infra/FMS2/MOM_domain_infra.F90:709
 3 0x000000000067957e __mom_open_boundary_MOD_open_boundary_init()  //import/c1/AKWATERS/kate/ESMG/ESMG-configs/src/MOM6/src/core/MOM_open_boundary.F90:1941
 4 0x0000000000c17638 __mom_state_initialization_MOD_mom_initialize_state()  //import/c1/AKWATERS/kate/ESMG/ESMG-configs/src/MOM6/src/initialization/MOM_state_initialization.F90:606
 5 0x0000000000924c7b __mom_MOD_initialize_mom()  //import/c1/AKWATERS/kate/ESMG/ESMG-configs/src/MOM6/src/core/MOM.F90:2979
 6 0x00000000005a2e70 MAIN__()  //import/c1/AKWATERS/kate/ESMG/ESMG-configs/src/MOM6/config_src/drivers/solo_driver/MOM_driver.F90:280
 7 0x000000000040761d main()  //import/c1/AKWATERS/kate/ESMG/ESMG-configs/src/MOM6/config_src/drivers/solo_driver/MOM_driver.F90:27
 8 0x000000000003a7e5 __libc_start_main()  ???:0
 9 0x000000000040765e _start()  ???:0
=================================

Program received signal SIGSEGV: Segmentation fault - invalid memory reference.
```

The line in question is:
```
  if (OBC%radiation_BCs_exist_globally) call pass_vector(OBC%rx_normal, OBC%ry_normal, G%Domain, &
                     To_All+Scalar_Pair)
```
At this point in the run, OBC%rx_normal and ry_normal have not been allocated - when the rotation is in play, itis allocated otherwise.

**Comment by kshedstrom** [comment]
In MOM_state_initialization, there is a CS%OBC which has the r[xy]_normal allocated. There is also an OBC which does not have them initialized (in the rotated case).

**Comment by kshedstrom** [comment]
Also, Thomas' test is sensitive to zeroing out the outside eta, but none of the rest of my tests are. It must be something about the OBC choices he picked. As for the other issue, perhaps @marshallward knows why the rotated MOM_state_initialization gets one version of the OBC structure and the nonrotated ones get the other?

**Comment by kshedstrom** [comment]
I ran dueling debuggers and the answers do match inside the domain, spinning up a gyre and all. The Thomas test has a positive eta_outside in the Flather OBC. That drives fluid from outside to the inside. It sucks water from the point just outside the boundary, causing eta to get lower just outside the boundary without the MOM_barotropic fix, eventually making it lower than the ocean bottom.

I still stand by my fix, except for the rotational weirdness.

**Comment by kshedstrom** [comment]
I'm trying a pointer fix to the OBC data structure problem. Things seem to be running...

**Comment by kshedstrom** [comment]
Running, maybe, but these are the diffs for the not rotated vs rotated Thomas test:
```
diff --git a/ocean_only/interior_OBCs/ocean.stats.gnu b/ocean_only/interior_OBCs/ocean.stats.gnu
index f7ee4fd9..fad32331 100644
--- a/ocean_only/interior_OBCs/ocean.stats.gnu
+++ b/ocean_only/interior_OBCs/ocean.stats.gnu
@@ -1,7 +1,7 @@
   Step,       Day,  Truncs,      Energy/Mass,      Maximum CFL,  Mean Sea Level,  Total Mass,  Mean Salin, Mean Temp, Frac Mass Err,   Salin Err,    Temp Err
             [days]                 [m2 s-2]           [Nondim]       [m]             [kg]         [PSU]      [degC]       [Nondim]        [PSU]        [degC]
      0,  726482.000,     0, En 1.2731932183656516E-25, CFL  0.00000, SL  5.0974E-13, M 1.40039E+14, S 17.5000, T 13.5000, Me  0.00E+00, Se  0.00E+00, Te  0.00E+00
-    36,  726482.250,     0, En 9.5246101764079154E-04, CFL  0.16364, SL  9.8685E-02, M 1.41421E+14, S 17.6380, T 13.4429, Me  9.77E-03, Se  3.09E-01, Te  7.48E-02
-    72,  726482.500,     0, En 1.7087457679639225E-03, CFL  0.19862, SL  9.6871E-02, M 1.41396E+14, S 17.8932, T 13.3365, Me -1.80E-04, Se  2.52E-01, Te -1.09E-01
-   108,  726482.750,     0, En 1.3284186657244862E-03, CFL  0.17970, SL  1.0015E-01, M 1.41442E+14, S 18.0322, T 13.2783, Me  3.24E-04, Se  1.45E-01, Te -5.39E-02
-   144,  726483.000,     0, En 1.1744820816381933E-03, CFL  0.17486, SL  1.0224E-01, M 1.41471E+14, S 18.0998, T 13.2499, Me  2.07E-04, Se  7.13E-02, Te -2.56E-02
+    36,  726482.250,     0, En 1.2731932183656516E-25, CFL  0.00000, SL  5.0974E-13, M 1.40039E+14, S 17.5000, T 13.5000, Me  0.00E+00, Se -1.95E-14, Te -5.74E-15
+    72,  726482.500,     0, En 1.2731932183656516E-25, CFL  0.00000, SL  5.0974E-13, M 1.40039E+14, S 17.5000, T 13.5000, Me  0.00E+00, Se -1.11E-14, Te -6.62E-15
+   108,  726482.750,     0, En 1.2731932183656516E-25, CFL  0.00000, SL  5.0974E-13, M 1.40039E+14, S 17.5000, T 13.5000, Me  0.00E+00, Se -1.48E-14, Te -8.96E-15
+   144,  726483.000,     0, En 1.2731932183656516E-25, CFL  0.00000, SL  5.0974E-13, M 1.40039E+14, S 17.5000, T 13.5000, Me  0.00E+00, Se -8.73E-15, Te -5.17E-15
```
Maybe I'll figure it out tomorrow.

**Comment by kshedstrom** [comment]
For the non-rotated case, OBC%segment(1)%field(1)%buffer_dst is allocated and set to the SSH outside value.
For the rotated case, it is not allocated. Then in MOM_barotropic, BT_OBC%SSH_outer_u is all zeroes instead of having the correct value at the boundaries.

**Comment by kshedstrom** [comment]
Oh, it got allocated, set to the appropriate value, then decallocated.

**Comment by kshedstrom** [comment]
Is this where it populated the whole OBC_in structure, then gets rid of it as part of the rotation? @marshallward

**Comment by kshedstrom** [comment]
rotate_OBC_segment_data is called after the segment fields with a value use the value to fill buffer_dst. It copies over the value, but not the filled buffer_dst.

**Comment by kshedstrom** [comment]
Something is going on with the tracer reservoirs. At some point during initialization, they have the same values. Later, before step_MOM_dynamics, they don't. In fact the "vanilla" case has updated to T=7 and the rotated case has not updated and has somehow picked up that is_initialized is .false..

**Comment by kshedstrom** [comment]
Summary of what I know about all this. Thomas Neumann asked first in the MOM6 forum, then in a MOM6 issue, about open boundaries within the domain. My tests for these were shown to be incomplete. I have managed to put in enough changes to make Dr Neumann happy, but the rotational tests for these things still have problems.

For the rotational tests of OBCs, the model initializes parts of CS%OBC, but simultaneously keeps OBC_in. It is this latter structure which gets passed to MOM_state_initialization. Bits are then copied over to CS%OBC, but not quite all of them in all cases.

1. The bits of the OBC structure which are saved on restart failed before I made them into pointers and made the pointers point to the CS%OBC bits.
2. In MOM_barotropic, the updates to all the interior points are fine, but there are non-zero updates to eta at h points just outside the OBC. If the run goes long enough, these points can have negative thicknesses. I have made it look more symmetric.
3. Fix the land masking for the OBCs when a masking depth is used.
4. Changed tr_reg to tr_Reg to match the rest.
5. Some experimentation with the rotate_OBC subroutines. Dr Neumann's test uses boundary data of the value=const type. Copying the buffer_dst from OBC_in to CS%OBC gets some of these across, but not the tracer reservoir values. I'm stumped.
6. Similar to the barotropic issue with eta outside, the tracer advection updates the tracer values outside. The flux values being used at the boundary cause these outside values to go out of bounds for the equation of state.

**Comment by kshedstrom** [comment]
I could be done with this for now. It's working without rotations as far as I know.

**Comment by Hallberg-NOAA** [comment]
This PR includes numerous commits that have nothing to do with it.  Please rebase this atop the latest version of dev/gfdl, which should hopefully eliminate some of the excess commits, most of which have already been merged into dev/gfdl.

File: src/core/MOM_barotropic.F90 [hunk 1]
```diff
@@ -2448,17 +2449,69 @@ subroutine btstep(U_in, V_in, eta_in, dt, bc_accel_u, bc_accel_v, forces, pbce,
                       haloshift=iev-ie, unscale=US%L_to_m**2*GV%H_to_m)
     endif
 
+    do j=jsv,jev
```
Comment (Hallberg-NOAA) [inline src/core/MOM_barotropic.F90:2452]: I see where this is going, but I have 4 specific problems with the changes to MOM_barotropic.F90:
1. This code setting `factor` does not change from one barotropic timestep to the next, and it might not even change between calls to btstep if the OBC segments don't change.  It should be moved before the barotropic time-stepping loop that starts at line 1815, and perhaps it should be made into an element of `barotropic_CS` and set in `barotropic_init()`.
2. Why isn't there a similar change from using `CS%IareaT` to `factor` where `eta_pred` is calculated on lines 1928, 1942 or 1948?
3. `factor` is not a very specific name that it does not reflect what this array does.  Something like `IareaT_OBCmask` would reflect what this array contains, even if it is somewhat less pithy than I would like.
4. The comment describing factor does not indicate that its units are `[L-2 ~> m-2]`.

File: src/core/MOM.F90 [hunk 2]
```diff
@@ -2897,6 +2897,26 @@ subroutine initialize_MOM(Time, Time_init, param_file, dirs, CS, &
     ! reservoirs are used.
     call open_boundary_register_restarts(HI, GV, US, CS%OBC, CS%tracer_Reg, &
                           param_file, restart_CSp, use_temperature)
+    if (turns /= 0) then
```
Comment (Hallberg-NOAA) [inline src/core/MOM.F90:2900]: We are getting segmentation faults in the TC testing from code related to the rotated tracer reservoirs, which means that  we can not accept this PR because all future PRs will also fail testing.

  I suspect that the problem might be that when there are is rotation by 1 or 3 turns, we should have `OBC_in%tres_x` corresponding to `CS%OBC%tres_y` and vice versa.  However, when there is rotation, these are (grid-space) arrays with different orientations and sizes so one can not simply point to the other.  There has to be an explicit copy instead.  These are scalars, so the sign will not change.

  The specific problem that is leading to a segmentation fault might be something as simple as the fact that the pointers are just not nullified when they are declared on line 372 of MOM_open_boundary.F90, or it could be that by setting a pointer to point to another pointer, the `associated` test appears to be true, even if the pointer that is pointed to is not itself pointing to allocated data.

  The same issues would be true of `OBC_in%rx_oblique_u` and the other fields, but these are only used for restarts, and I am pretty sure that these fields are never used from within `OBC_in`, so they might not even matter.

  Also, although the OBC type is not opaque, I think that we should treat it as though it were to the extent possible, so if it does still prove necessary to retain this code, I think that this whole block of code should be moved into a subroutine in MOM_open_boundary.F90 that copies these fields from one `ocean_OBC_type` to another.

File: src/core/MOM_open_boundary.F90 [hunk 3]
```diff
@@ -6130,6 +6147,93 @@ subroutine rotate_OBC_segment_data(segment_in, segment, turns)
...
+    do n = 1, segment_in%tr_Reg%ntseg
+      call rotate_array(segment_in%tr_Reg%tr(n)%tres, turns, segment%tr_Reg%tr(n)%tres)
+      ! Testing this to see if it works for contant tres values. Probably wrong otherwise.
+      segment%tr_Reg%Tr(n)%is_initialized=.true.
+    enddo
+  endif
+
+  ! Only doing this here for 1 turns, get flow directed in the opposite direction.
+  ! Currents that are now E_W were N_S and the turns should change their sign.
+  if (turns == 1 .and. segment%is_E_or_W) then
```
Comment (Hallberg-NOAA) [inline src/core/MOM_open_boundary.F90:6202]: I believe that perhaps the logic here should be:

> `qturns = modulo(turns,4)`
> `if ( (qturns == 2) .or. (qturns == 1 .and. segment%is_E_or_W) .or. (qturns == 3 .and. segment%is_N_or_S) ) then`

**Review by Hallberg-NOAA — CHANGES_REQUESTED** [review]
Thank you very much for doing the rebase, @kshedstrom .  That makes it so much easier to see what is going on with this PR!

  I have added some questions and thoughts about what might be going on here and some things to try that might fix our testing failures, if you are feeling up to it.

Another option here might be to break up this PR into the first few commits that we can accept without controversy (namely those related to setting MASKING_DEPTH for the OBCs and masking out the tracer-point changes in the barotropic solver and tracer advection), and then circling back to sort out what is happening with the rotation later.  (The commit changing the allocatable arrays into pointers seems to me like the primary suspect for the TC testing failures, based on the error messages we are getting with our TC testing failures.  To see the errors, click on the red-x's that we are getting at https://github.com/NOAA-GFDL/MOM6/pull/752.)

File: src/core/MOM_barotropic.F90 (hunk 1 above)
Comment (kshedstrom) [reply src/core/MOM_barotropic.F90:2452]: You have some good points about the factor code. The BT_OBC structure is set up for each call to btstep, so if we don't fix that, it will be hard to move anything to barotropic_init().

**Comment by kshedstrom** [comment]
Good idea to try without the rotational fixes for now.

**Comment by Hallberg-NOAA** [comment]
@kshedstrom, I think that we have made significant progress toward addressing some of the issues that this PR was intended to address working via PR #814, but this has now created conflicts with this PR so we can not simply rebase it.  Could I ask that you please rebase this PR so that we can work out how to handle the other aspects of what it is striving for?

**Comment by Hallberg-NOAA** [comment]
Please rebase this PR on top of dev/gfdl instead of doing a merge commit, @kshedstrom.  With the merge commit, this PR appears to have 57 commits, whereas with a properly executed rebase it would have far fewer, and it would set us up to be able to merge this commit onto dev/gfdl in due course while maintaining a linear commit history.

File: src/core/MOM.F90 (hunk 2 above)
Comment (Hallberg-NOAA) [reply src/core/MOM.F90:3059]: I think that this block of code still needs to have an alternative that swaps the roles of the `..._u` and `..._v` components when we are rotating by 90 or 270 degrees, but otherwise I think that this PR could be ready to go.

File: src/core/MOM_open_boundary.F90 [hunk 4]
```diff
@@ -6145,6 +6156,56 @@ subroutine rotate_OBC_segment_data(segment_in, segment, turns)
...
+       call rotate_array(segment_in%tangential_vel, turns, segment%tangential_vel)
+  if (allocated(segment_in%tangential_grad)) &
+       call rotate_array(segment_in%tangential_grad, turns, segment%tangential_grad)
+  if (allocated(segment_in%grad_normal)) &
+       call rotate_array(segment_in%grad_normal, turns, segment%grad_normal)
+  if (allocated(segment_in%grad_tan)) &
+       call rotate_array(segment_in%grad_tan, turns, segment%grad_tan)
+  if (allocated(segment_in%grad_gradient)) &
+       call rotate_array(segment_in%grad_gradient, turns, segment%grad_gradient)
+  if (allocated(segment_in%rx_norm_rad)) &
```
Comment (Hallberg-NOAA) [inline src/core/MOM_open_boundary.F90:6185]: The call on the following line that is going to copy the x- norm_rad array of the input segment into the y- norm_rad of the rotated component should work when there are 1 or 3 turns (90 or 270 degrees), but I don't see how this would be the right thing to do when we are rotating by 2 turns (180 degrees).  The same consideration would apply to lines 6188, 6190 and 6192.

File: src/core/MOM_open_boundary.F90 [hunk 5]
```diff
@@ -6145,6 +6156,102 @@ subroutine rotate_OBC_segment_data(segment_in, segment, turns)
...
+        if (allocated(segment%nudged_normal_vel)) &
+            segment%nudged_normal_vel(:,:,:) = -segment%nudged_normal_vel(:,:,:)
+      else
+        if (allocated(segment%tangential_vel)) &
+            segment%tangential_vel(:,:,:) = -segment%tangential_vel(:,:,:)
+        if (allocated(segment%nudged_tangential_vel)) &
+            segment%nudged_tangential_vel(:,:,:) = -segment%nudged_tangential_vel(:,:,:)
+      endif
+    elseif ((segment%field(n)%name == 'V' .or. segment%field(n)%name == 'Vamp') .and. &
+         modulo(turns, 4) == 3) then
```
Comment (Hallberg-NOAA) [inline src/core/MOM_open_boundary.F90:6238]: There is an asymmetry between lines 6223 and 2238 that I do not understand.  For a u-field, the sign of `buffer_dst` and the other fields is flipped for rotations or 1 or 3 turns, but for a v-field it is only flipped if there are 3 turns.  Logically, it seems like there should be the same number of cases where the signs are flipped for both u and v fields.

**Comment by kshedstrom** [comment]
I believe I have fixed your specific issue. However, at line #2886 of MOM.F90, there is a fatal error when attempting turns other than 1 with OBCs. Commenting this out causes my test for turns=2 to fail elsewhere and turns=3 to run but not look correct. I can pursue these later.

File: src/core/MOM_open_boundary.F90 [hunk 6]
```diff
@@ -2016,16 +2018,16 @@ subroutine open_boundary_dealloc(OBC)
...
-  if (allocated(OBC%ry_normal)) deallocate(OBC%ry_normal)
-  if (allocated(OBC%rx_oblique_u)) deallocate(OBC%rx_oblique_u)
-  if (allocated(OBC%ry_oblique_u)) deallocate(OBC%ry_oblique_u)
-  if (allocated(OBC%rx_oblique_v)) deallocate(OBC%rx_oblique_v)
-  if (allocated(OBC%ry_oblique_v)) deallocate(OBC%ry_oblique_v)
-  if (allocated(OBC%cff_normal_u)) deallocate(OBC%cff_normal_u)
-  if (allocated(OBC%cff_normal_v)) deallocate(OBC%cff_normal_v)
-  if (allocated(OBC%tres_x)) deallocate(OBC%tres_x)
-  if (allocated(OBC%tres_y)) deallocate(OBC%tres_y)
+  if (associated(OBC%rx_normal)) nullify(OBC%rx_normal)
```
Comment (Hallberg-NOAA) [inline src/core/MOM_open_boundary.F90:2021]: Even though these allocatable arrays have been replaced with pointers, I think that we still need to deallocate them, rather than nullifying them, to avoid a memory leak.  This potential memory leak may not matter if we only call `open_boundary_deallocate()` at the end of a run.

Comment (kshedstrom) [reply src/core/MOM_open_boundary.F90:2032]: Good suggestion to clean this back up. Done.

File: src/core/MOM.F90 (hunk 2 above)
Comment (Hallberg-NOAA) [reply src/core/MOM.F90:3097]: This issue does not break anything that is working yet, so I think that we can let this go for now. We will deal with the issues related to rotation by 90 or 270 degrees in subsequent PRs.

Comment (Hallberg-NOAA) [reply src/core/MOM.F90:3097]: The TC testing is no longer having these problems with segmentation faults, so I think that we can accept this PR as it stands and correct other issues related to grid rotation in subsequent PRs.

File: src/core/MOM_open_boundary.F90 (hunk 5 above)
Comment (Hallberg-NOAA) [reply src/core/MOM_open_boundary.F90]: I agree that this asymmetry has now been corrected.

**Review by Hallberg-NOAA — APPROVED** [review]
Thank you for this contribution.   I am now satisfied that this is an important step forward and can be accepted, even if there are still some remaining issues related to grid rotation with open boundary conditions.

**Comment by Hallberg-NOAA** [comment]
This PR has passed pipeline testing at https://gitlab.gfdl.noaa.gov/ogrp/mom6ci/MOM6/-/pipelines/27367.

**Review by adcroft — APPROVED** [review]
